### ISSUE TITLE
correct way to make loop with promise in nodejs

### ISSUE THREAD

**Issue by AnD010** [kriskowal/q#770]
Hi, I hope someone can help me better understand the promises.
I'm pretty rookie with them.
I'm doing a DB query with promises, when the result is correct get an array of data. Then I need to make a loop iterating this array and perform for each element delete operation and after that delete operation to another operation of sending, when both operations finished, then,  repeat the process for each element of the array, but must be completed for each element the finish operation and send operation before starting another array element ... with my code that does not happen, the loop make delete operation for each element without awaiting the delete and send operation return result....
 Both operation delete and send have the structure

    function x (makeQuery){
      .......
      var deferred = Q.defer();
       .....
       function(err, result) {
                        if (err) {
                          deferred.reject(err);
                        } else {
                         deferred.resolve(result);
                        }
                        
                    });
      return deferred.promise;
    }
And my code is :

     getElementInPostgres(makeQuery)
            .then(function (obj) {
              if (...){
                    results= obj['rows']
                    index =0
                    for (var i = 0; i < results.length; i++) {
                        var notification = {}
                      notification.tag='SubscriptionNode'
                      notification.indexSocket=results[i]['indexsocket']
                      notification.clientID=results[i]['clientid']
                      notification.callbackURL=results[i]['callbackurl']
                      notification.deploymentID=results[i]['deploymentid']
                      notification.clusterID=results[i]['clusterid']
                      notification.date=results[i]['registered']
                      var query = "DELETE FROM.....";
                      var params = [notification.indexSocket];
                      deleteInPostgres(query,params,notification)
                      .then(function (notification) {
                          if(notification.clientID){
                              sendInfoToPython(notification)
                               .then(function (obj) {
                                   ..........
                               }).fail(function (err) {
                                  ......
                               });
                          }
                    
                      }).fail(function (err) {
                        ......
                       });
                     }
                   }
                   else{
                    .........
                   }
                  }).fail(function (err) {
                   
                    });

**Comment by gitoverflow** [comment]
```
var daoFunction = function () {

  var rootDeffered = Q.defer();

  var getData = function () {

    var deferred = Q.defer();
    // your async stuff

    setTimeout(function () {

      var data = [];

      var i = 0;
      while (i < 10) {
        data.push(i);
        i++
      }

      console.log("data fetched");

      //resolved when data is fetched
      deferred.resolve(data);
    }, 3000);

    // return promise
    return deferred.promise;

  };

  var deleteItem = function (data, index) {

    // call sendFunction after deleteItem finishes
    sendFunction(data, index);

  };

  var sendFunction = function (data, index) {

    // call this, when your sendFunction is done
    if (data.length !== index + 1) {
      deleteItem(data, index + 1)
    }

    if (data.length === index + 1) {

      rootDeffered.resolve(data);

    }

  };

  getData()
    .then(function (data) {

      deleteItem(data, 0);

    })
    .catch(function (err) {
      console.log(err)
    });

  return rootDeffered.promise;

};

daoFunction()
  .then(function (data) {
    console.log(data)
  })
  .catch(function (err) {
    // whatever
  });

```

**Comment by panzerdp** [comment]
@AnD010 Post your question to stackoverflow. This section is more for issues/tasks/planning/discussions.

**Comment by AnD010** [comment]
@gitoverflow thanks so much for you answer .
@panzerdp The next time,  I will write my problem in stackoverflow , thanks
